// fetch-logistics/dist/dispatcher.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Dispatcher = void 0;
const rxjs_1 = require("rxjs");
const redis_observable_1 = require("redis-observable");
let Dispatcher = (() => {
    class Dispatcher {
        constructor() {
            this.chose = (param, handler_name = null) => {
                for (const handler of this.handlers) {
                    if (handler.legal(param) &&
                        (handler_name ? handler.name === handler_name : true)) {
                        return handler;
                    }
                }
                return null;
            };
            this.query = (param, handler_name = null, force = true) => {
                const handler = this.chose(param, handler_name);
                return (handler
                    ? (force ? handler.query : handler.getCacheOrInit)(param, force)
                    : rxjs_1.of(null));
            };
            this.getHandlerName = () => {
                return this.handlers.map(i => i.name);
            };
            this.getHandler = (handler_name) => {
                for (const handler of this.handlers) {
                    if (handler.name === handler_name) {
                        return handler;
                    }
                }
                return null;
            };
            this.getConfig = () => {
                return Object.fromEntries(this.handlers.map(i => [i.name, i.getConfig()]));
            };
            this.setConfig = (handler_name, option) => {
                const handler = this.getHandler(handler_name);
                if (handler) {
                    handler.setConfig(option);
                    return handler.getConfig()[handler.name];
                }
                return null;
            };
        }
    }
    Dispatcher.register = (creator, config) => {
        if (!Dispatcher.instance) {
            throw new Error('Must create a Dispatcher instance');
        }
        for (const handler of Dispatcher.instance.handlers) {
            if (handler.name === config.name) {
                handler.setConfig(config);
                return;
            }
        }
        Dispatcher.instance.handlers.push(new creator({ redis: Dispatcher.redis, config }));
        Dispatcher.instance.handlers = Dispatcher.instance.handlers.sort((a, b) => a.weight - b.weight);
        return;
    };
    Dispatcher.create = (redisOpt) => {
        if (Dispatcher.instance) {
            return Dispatcher.instance;
        }
        if (!Dispatcher.redis) {
            Dispatcher.redis = new redis_observable_1.RxRedis(redisOpt);
        }
        Dispatcher.instance = new Dispatcher();
        return Dispatcher.instance;
    };
    return Dispatcher;
})();
exports.Dispatcher = Dispatcher;
