// @jsdvjx/query-fun/dist/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.QueryFun = void 0;
const string_fun_1 = require("@jsdvjx/string-fun");
const fs_1 = require("fs");
const YAML = require("yamljs");
class QueryFun {
    constructor(options) {
        this.defOption = { singleRow: false, tag: 'default', autoType: true };
        this.getParams = () => {
            return this.strFun.getDefaultParam();
        };
        this.initQuerier = () => {
            return this.options.reduce((res, option) => {
                res[option.name] = (param) => {
                    const sql = this.strFun.caller[option.name](param);
                    return (QueryFun.queries.get(option.tag) ||
                        QueryFun.queries.get('default') || {
                        query: (_p) => {
                            return Promise.reject(`Querier not found, param=${JSON.stringify(_p)} `);
                        },
                    })
                        .query(sql)
                        .then(QueryFun.createHandler(option));
                };
                return res;
            }, {});
        };
        this.demo = async () => {
            const result = {};
            const params = this.getParams();
            const map = {
                string: '',
                number: 1,
                date: new Date(),
                boolean: true,
                bigint: 1n,
            };
            for (const option of this.options) {
                const _p = params[option.name];
                if (!option.autoType && option.type) {
                    result[option.name] = map[option.type];
                }
                if (option.autoType) {
                    if (_p)
                        result[option.name] = await this.querier[option.name](_p);
                }
            }
            return result;
        };
        this.strDemo = () => {
            return this.strFun.demo;
        };
        this.close = () => {
            for (const iterator of QueryFun.queries.values()) {
                iterator.close();
            }
        };
        this.exportInterface = async () => {
            const Result = await this.demo();
            const Input = this.strDemo();
            return {
                Result,
                Input,
            };
        };
        this.options = options.map((i) => ({
            ...this.defOption,
            ...i,
        }));
        this.querier = this.initQuerier();
        this.strFun = string_fun_1.StringFun.createByTemplate(this.options);
    }
}
exports.QueryFun = QueryFun;
QueryFun.queries = new Map();
QueryFun.register = async (querist) => {
    if (!querist)
        return;
    try {
        await querist.init();
        QueryFun.queries.set(querist.tag, querist);
        if (!QueryFun.queries.has('default')) {
            QueryFun.queries.set('default', querist);
        }
    }
    catch (e) {
        return;
    }
};
QueryFun.create = (path, format = 'yaml') => {
    if (QueryFun.instance) {
        return QueryFun.instance;
    }
    try {
        return ((QueryFun.instance = new QueryFun((fs_1.readdirSync(path) || [])
            .map((i) => {
            const _path = `${path}/${i}`;
            return (format === 'json' ? JSON : YAML).parse(fs_1.statSync(_path).isFile()
                ? fs_1.readFileSync(_path).toString()
                : '[]');
        })
            .reduce((res, acc) => {
            return res.concat(acc);
        }, []) || [])),
            QueryFun.instance);
    }
    catch (e) {
        console.log(e);
    }
};
QueryFun.query = (sql, tag = 'default') => {
    return QueryFun.queries.get(tag).query(sql);
};
QueryFun.handler = {};
QueryFun.createHandler = (option) => {
    const key = `${option.tag}.${option.name}`;
    return QueryFun.handler[key]
        ? QueryFun.handler[key]
        : ((QueryFun.handler[key] = (result) => {
            let _result = result;
            if (option.singleRow) {
                _result = result[0];
                if (option.pickField) {
                    _result =
                        _result[option.pickField] || Object.values(_result).shift();
                    if (option.type) {
                        switch (option.type) {
                            case 'bigint':
                                _result = BigInt(_result);
                                break;
                            case 'number':
                                _result = /\d+\.\d+/.test(_result)
                                    ? parseFloat(_result)
                                    : parseInt(_result);
                                break;
                            case 'date':
                                _result = new Date(_result);
                                break;
                            case 'string':
                            default:
                                break;
                        }
                    }
                }
            }
            return _result;
        }),
            QueryFun.handler[key]);
};
