// fetch-logistics/dist/express.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.IExpress = exports.ExpressState = void 0;
const rxjs_1 = require("rxjs");
const operators_1 = require("rxjs/operators");
const crypto = require("crypto");
const levenshtien = require("damerau-levenshtein");
const dayjs = require("dayjs");
var ExpressState;
(function (ExpressState) {
    ExpressState[ExpressState["UNACTIVE"] = -1] = "UNACTIVE";
    ExpressState[ExpressState["NOTFOUND"] = 0] = "NOTFOUND";
    ExpressState[ExpressState["TRANSIT"] = 1] = "TRANSIT";
    ExpressState[ExpressState["PICKUP"] = 2] = "PICKUP";
    ExpressState[ExpressState["DELIVERED"] = 3] = "DELIVERED";
    ExpressState[ExpressState["EXCEPTION"] = 4] = "EXCEPTION";
    ExpressState[ExpressState["EXPIRED"] = 5] = "EXPIRED";
})(ExpressState = exports.ExpressState || (exports.ExpressState = {}));
class IExpress {
    constructor(option) {
        this.expire = 3600 * 24 * 10;
        this.setConfig = (config) => {
            this.config = this.config || config.conf;
            this.webhook = config.webhook || this.webhook;
            this.weight = config.weight || this.weight;
            this.rate = config.rate || this.rate;
            this.max_count = config.max_count || this.max_count;
            this.checkList = config.policy ||
                this.checkList || { type: 'none', codes: [] };
            this.diy_map = config.diy_map;
            this.expire = config.expire || this.expire;
        };
        this.getConfig = () => {
            return {
                conf: this.config,
                webhook: this.webhook,
                weight: this.weight,
                rate: this.rate,
                max_count: this.max_count,
                policy: this.checkList,
                signTemplates: this.signTemplates,
                expire: this.expire,
            };
        };
        this.webhook = false;
        this.weight = 100;
        this.rate = 0;
        this.max_count = 200;
        this.legal = (param) => {
            if (!this.check(param.code)) {
                return false;
            }
            const code = this.fixCode(param).code;
            return this.codes.has(code) && this.check(code);
        };
        this.check = (code) => {
            switch (this.checkList.type) {
                case 'white':
                    return this.checkList.codes.includes(code);
                case 'black':
                    return !this.checkList.codes.includes(code);
                case 'none':
                default:
                    return true;
            }
        };
        this.put = (param) => {
            return this.getCacheOrInit(param).pipe(operators_1.mergeMap((info) => {
                if (info.request_count > 0) {
                    return rxjs_1.of(info);
                }
                return this._put(info).pipe(operators_1.tap((i) => this.after_convert([i, i])));
            }));
        };
        this.query = (param, force = false) => {
            if (!this.checkExNu(param)) {
                throw new Error(`${param.company},${param.number} illegal`);
            }
            return this.getCacheOrInit(this.fixCode(param)).pipe(operators_1.mergeMap((info) => {
                console.log(JSON.stringify(info), 'QUERY_FL');
                if (info.state === ExpressState.DELIVERED) {
                    return rxjs_1.of(info);
                }
                if (info.request_count >= this.max_count && !force) {
                    return rxjs_1.of(info);
                }
                if (this.rate > Date.now() / 1000 - (info.last_request || 0)) {
                    return rxjs_1.of(info);
                }
                if (this.webhook && !force) {
                    return rxjs_1.of(this.after_convert(this.convert([info, info.source]))[0]);
                }
                console.log('REAL_QUERY', info.number, info.company);
                return this.fetch(info).pipe(operators_1.mergeMap((source) => this.push(source, param)));
            }));
        };
        this.push = (source, param) => {
            return rxjs_1.zip(this.getCacheOrInit(param ? param : this.getParamByResponse(source)), rxjs_1.of(source)).pipe(operators_1.map(this.convert), operators_1.tap(this.after_convert), operators_1.map(([result, _]) => result));
        };
        this.after_convert = ([last, pre]) => {
            last.request_count += 1;
            last.updated_at = new Date();
            last.last_request = Math.floor(last.updated_at.valueOf() / 1000);
            if (pre.state !== ExpressState.DELIVERED) {
                this.redis
                    .set(this.getKey(last), JSON.stringify(last), 3600 * 24 * 10)
                    .subscribe();
            }
        };
        this.expired = (last_request) => {
            return last_request + this.rate <= Math.floor(Date.now() / 1000);
        };
        this.checkExNu = (param) => {
            const nu = param.number;
            const reg = /^[a-zA-Z0-9]{6,16}$/;
            const result = reg.test(nu);
            if (!result)
                console.error(`${param.company},${param.number}`, 'illegal');
            return result;
        };
        this.initExpressInfo = (param) => {
            const now = new Date();
            const result = {
                ...param,
                state: ExpressState.UNACTIVE,
                source_state: ExpressState.UNACTIVE,
                guess: false,
                data: [],
                source: null,
                last_request: 0,
                type: this.name,
                md5: null,
                request_count: 0,
                created_at: now,
                updated_at: now,
            };
            return result;
        };
        this.getCacheOrInit = (param) => {
            const key = this.getKey(param);
            return this.redis.get(key).pipe(operators_1.mergeMap((cache) => {
                if (cache === null) {
                    return rxjs_1.of(this.initExpressInfo(param));
                }
                else {
                    return rxjs_1.of(JSON.parse(cache));
                }
            }));
        };
        this.getKey = (param) => {
            return `EXPRESS_CACHE_${this.name}_${param.code}_${param.number}`;
        };
        this.convert = ([container, source]) => {
            const _state = this.getState(source);
            const data = this.getProcess(source).sort((a, b) => dayjs(b.time).valueOf() - dayjs(a.time).valueOf());
            const guess = this.guess_sign(data, container.code);
            const request_count = container.request_count;
            const now = new Date();
            const state = _state !== ExpressState.DELIVERED && guess
                ? ExpressState.DELIVERED
                : _state;
            const md5 = this.getMd5(data, state);
            const source_state = _state;
            return [
                {
                    ...container,
                    data,
                    state,
                    source_state,
                    guess,
                    source,
                    md5,
                    request_count,
                    updated_at: now,
                },
                container,
            ];
        };
        this.getMd5 = (last, state) => {
            return crypto
                .createHash('md5')
                .update(last.map((i) => i.time.toISOString()).join('') + '_' + state)
                .digest('hex');
        };
        this.levenshtien = levenshtien;
        this.guess_sign = (process, code) => {
            const lastInfo = process[0];
            if (!lastInfo) {
                return false;
            }
            if (dayjs(lastInfo.time).add(1, 'd').isAfter(dayjs())) {
                return false;
            }
            for (const template of this.signTemplates) {
                if (template.code === code) {
                    for (const reg of template.regex) {
                        if (reg.test(lastInfo.content)) {
                            return true;
                        }
                    }
                }
            }
            return false;
        };
        this.codeMap = null;
        this.init = () => {
            return (this._init || (() => rxjs_1.of(null)))().pipe(operators_1.mergeMap(this.initCode), operators_1.tap((codeMap) => {
                this.codeMap = codeMap;
                this.codes = new Set(this.codeMap.map((i) => i.code));
            }), operators_1.map(() => this));
        };
        this.fixCode = (param) => {
            if (this.codes.has(param.code))
                return param;
            for (const i of this.diy_map || []) {
                if (param.company === i.company) {
                    param.code = i.code;
                    return param;
                }
            }
            const code = this.codeMap
                .map((tmp) => {
                const step = this.levenshtien(tmp.code, param.code || '').steps +
                    this.levenshtien(tmp.company, param.company).steps;
                return { ...tmp, step };
            })
                .sort((a, b) => b.step - a.step)
                .pop();
            return { ...param, ...code };
        };
        this.redis = option.redis;
        this.setConfig(option.config);
    }
}
exports.IExpress = IExpress;
