// fetch-logistics/dist/handler/ali.handler.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AliHandler = void 0;
const express_1 = require("../express");
const rxjs_1 = require("rxjs");
const querystring = require("querystring");
const axios_1 = require("axios");
const operators_1 = require("rxjs/operators");
let AliHandler = (() => {
    class AliHandler extends express_1.IExpress {
        constructor() {
            super(...arguments);
            this.expire = 3600 * 24 * 10;
            this.name = 'ALI';
            this.webhook = false;
            this.weight = 100;
            this.rate = 3600;
            this.max_count = 120;
            this.getState = data => parseInt(data.state) || express_1.ExpressState.UNACTIVE;
            this.getProcess = input => (input.list || []).map(i => ({
                time: new Date(i.time),
                content: i.content,
            }));
            this.legal = param => {
                const result = ['顺丰快递', '极兔速递'].includes(param.company);
                console.log(JSON.stringify(result), 'ALI_HANDLER');
                return result;
            };
            this.baseUrl = 'https://cexpress.market.alicloudapi.com/';
            this.fetch = (param) => {
                const task = param;
                const url = `${this.baseUrl}cexpress?${querystring.stringify({
                    no: task.code === 'SF'
                        ? `${task.number}:${task.phone.substr(task.phone.length - 4, 4)}`
                        : task.number,
                    type: task.code,
                })}`;
                return this.request(url);
            };
            this.request = (url, header = {}) => {
                return rxjs_1.from(axios_1.default
                    .get(url, {
                    responseType: 'text',
                    headers: {
                        ...header,
                        Authorization: `APPCODE ${this.config.appCode}`,
                    },
                })
                    .then(i => i.data));
            };
            this.signTemplates = [];
            this.initCode = () => {
                const url = `${this.baseUrl}cExpressLists`;
                return this.request(url).pipe(operators_1.map(tmp => tmp.result), operators_1.map((items) => {
                    return Object.entries(items).map(([code, company]) => ({
                        code,
                        company,
                    }));
                }));
            };
        }
    }
    AliHandler.getCreator = () => { };
    return AliHandler;
})();
exports.AliHandler = AliHandler;
