// fetch-logistics/dist/handler/tm.handler.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TmHandler = void 0;
const express_1 = require("../express");
const trackingmore_1 = require("trackingmore");
const rxjs_1 = require("rxjs");
const operators_1 = require("rxjs/operators");
class TmHandler extends express_1.IExpress {
    constructor() {
        super(...arguments);
        this._put = param => {
            return this.tm
                .createTracking({
                tracking_number: param.number,
                carrier_code: param.code,
            })
                .pipe(operators_1.mergeMap(i => rxjs_1.of(param)));
        };
        this.expire = 3600 * 24 * 10;
        this._init = () => {
            this.tm = new trackingmore_1.TrackingMoreApi(this.config.apiKey);
            return rxjs_1.of(this.tm);
        };
        this.name = 'TM';
        this.weight = 0;
        this.rate = 0;
        this.max_count = 1;
        this.getState = task => {
            switch (task.status) {
                case 'delivered':
                    return express_1.ExpressState.DELIVERED;
                case 'notfound':
                case 'pending':
                    return express_1.ExpressState.NOTFOUND;
                case 'pickup':
                    return express_1.ExpressState.PICKUP;
                case 'transit':
                    return express_1.ExpressState.TRANSIT;
                case 'exception':
                    return express_1.ExpressState.EXCEPTION;
                case 'undelivered':
                case 'expired':
                default:
                    return express_1.ExpressState.EXPIRED;
            }
        };
        this.getProcess = item => {
            if (!item || !item.origin_info || !item.origin_info.trackinfo) {
                return [];
            }
            return item.origin_info.trackinfo.map(item => ({
                time: new Date(item.Date),
                content: item.StatusDescription,
            }));
        };
        this.fetch = param => {
            const result = () => this.tm
                .getTracking({
                tracking_number: param.number,
                carrier_code: param.code,
            })
                .pipe(operators_1.map(i => {
                return i.data.data;
            }), operators_1.map(this.info2body));
            if (param.request_count === 0) {
                this.tm
                    .createTracking({
                    tracking_number: param.number,
                    carrier_code: param.code,
                })
                    .pipe(operators_1.mergeMap(() => result()));
            }
            return result();
        };
        this.info2body = info => {
            return info
                ? {
                    ...info,
                    customer_email: (info.customer_email || [null])[0],
                }
                : null;
        };
        this.signTemplates = [
            {
                regex: [/派件送达【.+?】.+?，请前往.+?领取您的包裹，联系电话：[\d-]+/],
                code: 'bestex',
            },
            {
                regex: [
                    /快件已在 【.+?】 签收, 签收人:.+? 如有疑问请电联:.+?, 您的快递已经妥投。风里来雨里去, 只为客官您满意。上有老下有小, 赏个好评好不好？【请在评价快递员处帮忙点亮五颗星星哦~】/,
                    /快件已送达【.+?】, 如有问题请电联.+?感谢.*?使用中通快递, 期待再次为您服务!/,
                ],
                code: 'zto',
            },
            {
                regex: [
                    /【.+?】.+? 快递员 .+?\d+? 正在为您派件【.+?为韵达快递员外呼专属号码，请放心接听】/,
                ],
                code: 'yunda',
            },
            {
                regex: [/快件已[由被].+?.+?，请及时取件[。，].+请联系\d+/],
                code: 'sto',
            },
            {
                regex: [/邮件投递到.+?,投递员:.+?,电话:.+?/],
                code: 'ems',
            },
        ];
        this.getParamByResponse = source => {
            for (const code of this.codeMap) {
                if (code.code === source.carrier_code) {
                    return {
                        id: -1,
                        company: code.company,
                        number: source.tracking_number,
                        code: code.code,
                        phone: '',
                        delivery_time: new Date(source.created_at),
                    };
                }
            }
            return null;
        };
        this.initCode = () => {
            return this.tm.carriers().pipe(operators_1.pluck('data'), operators_1.map(i => {
                return i.data.map(item => ({ company: item.name, code: item.code }));
            }));
        };
    }
}
exports.TmHandler = TmHandler;
