// fetch-logistics/dist/handler/tm.handler.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TmHandler = void 0;
const express_1 = require("../express");
const trackingmore_1 = require("trackingmore");
const rxjs_1 = require("rxjs");
const operators_1 = require("rxjs/operators");
class TmHandler extends express_1.IExpress {
    constructor() {
        super(...arguments);
        this.expire = 3600 * 24 * 10;
        this.checkList = {
            type: 'black',
            codes: ['sf-express'],
        };
        this._init = () => {
            this.tm = new trackingmore_1.TrackingMoreApi(this.config.apiKey);
            return rxjs_1.of(this.tm);
        };
        this.name = 'TM';
        this.weight = 0;
        this.rate = 0;
        this.max_count = 1;
        this.getState = task => {
            switch (task.status) {
                case 'delivered':
                    return express_1.ExpressState.DELIVERED;
                case 'notfound':
                case 'pending':
                    return express_1.ExpressState.NOTFOUND;
                case 'pickup':
                    return express_1.ExpressState.PICKUP;
                case 'transit':
                    return express_1.ExpressState.TRANSIT;
                case 'exception':
                    return express_1.ExpressState.EXCEPTION;
                case 'undelivered':
                case 'expired':
                default:
                    return express_1.ExpressState.EXPIRED;
            }
        };
        this.getProcess = item => {
            if (!item || !item.origin_info || !item.origin_info.trackinfo) {
                return [];
            }
            return item.origin_info.trackinfo.map(item => ({
                time: new Date(item.Date),
                content: item.StatusDescription,
            }));
        };
        this.fetch = param => {
            return this.tm
                .getTracking({
                tracking_number: param.number,
                carrier_code: param.code,
            })
                .pipe(operators_1.map(i => {
                return i.data.data;
            }), operators_1.map(this.info2body));
        };
        this.info2body = info => {
            return info
                ? {
                    ...info,
                    customer_email: info.customer_email[0],
                }
                : null;
        };
        this.signTemplates = [];
        this.getParamByResponse = source => {
            for (const code of this.codeMap) {
                if (code.code === source.carrier_code) {
                    return {
                        id: -1,
                        company: code.company,
                        number: source.tracking_number,
                        code: code.code,
                        phone: '',
                        delivery_time: new Date(source.created_at),
                    };
                }
            }
            return null;
        };
        this.initCode = () => {
            return this.tm.carriers().pipe(operators_1.pluck('data'), operators_1.map(i => {
                return i.data.map(item => ({ company: item.name, code: item.code }));
            }));
        };
    }
}
exports.TmHandler = TmHandler;
