// @jsdvjx/query-fun/dist/Queries/mysql.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const mysql = require("mysql2");
const IQuerist_1 = require("./IQuerist");
class MysqlQuerist extends IQuerist_1.IQuerist {
    constructor() {
        super(...arguments);
        this.close = () => {
            this.client.destroy();
        };
        this.query = (sql) => {
            const result = [];
            return new Promise((resolve, reject) => {
                this.client
                    .query(sql)
                    .on('result', (row, idx) => {
                    //@ts-ignore;
                    if (row.constructor.name !== 'TextRow') {
                        result.push(row.affectedRows || 0);
                    }
                    else {
                        result.push(row);
                    }
                })
                    .on('error', (e) => (console.error(sql, 'SQL_ERROR'), reject(e)))
                    .on('end', () => resolve(result));
            });
        };
        this.type = 'MYSQL';
        this.init = () => {
            return new Promise((resolve, reject) => {
                const option = this.option;
                delete option.tag;
                delete option.type;
                this.client = mysql.createConnection(this.option);
                this.client.connect((e) => (e ? reject(e) : resolve()));
            });
        };
    }
}
exports.default = MysqlQuerist;
