// @jsdvjx/query-fun/dist/Queries/postgresql.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const pg_1 = require("pg");
const IQuerist_1 = require("./IQuerist");
class PgQuerist extends IQuerist_1.IQuerist {
    constructor(option) {
        super(option);
        this.init = async () => {
            if (!this.inited) {
                this.inited = true;
                await this.client.connect();
                await this.client.query(`set search_path to ${(this.option.path || ['public'])
                    .map(i => `"${i}"`)
                    .join(',')}`);
            }
        };
        this.inited = false;
        this.query = async (sql) => {
            return this.client.query(sql).then(result => {
                switch (result.command) {
                    case 'SELECT':
                        return result.rows;
                    case 'DO':
                        return null;
                    case 'UPDATE':
                    case 'INSERT':
                        return result.rowCount;
                }
            });
        };
        this.close = async () => {
            await this.client.end();
        };
        this.type = 'POSTGRESQL';
        this.client = new pg_1.Client(option);
    }
}
exports.default = PgQuerist;
