// @jsdvjx/query-fun/dist/Queries/sls.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const sls_client_1 = require("sls-client");
const moment = require("moment");
const IQuerist_1 = require("./IQuerist");
class SlsQuerist extends IQuerist_1.IQuerist {
    constructor(option) {
        super(option);
        this.init = () => Promise.resolve();
        this.type = 'SLS';
        this.getQueryOption = (sql) => {
            const [from, to, filter, _sql] = sql.split('|');
            const f = 'YYYY-MM-DD HH:mm:ss';
            return {
                from: moment(from, f).unix(),
                to: (to === 'now' ? moment() : moment(to, f)).unix(),
                query: `${filter}|${_sql}`,
            };
        };
        this.close = () => { };
        this.query = sql => {
            const params = this.getQueryOption(sql);
            return this.client.getLogs(params).toPromise();
        };
        this.tag = option.tag;
        this.client = new sls_client_1.default(option);
    }
}
exports.default = SlsQuerist;
