// trackingmore/dist/index.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (Object.hasOwnProperty.call(mod, k)) result[k] = mod[k];
    result["default"] = mod;
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
// eslint-disable-next-line no-unused-vars
var axios_1 = __importDefault(require("axios"));
var rxjs_1 = require("rxjs");
var operators_1 = require("rxjs/operators");
var qs = __importStar(require("querystring"));
var fs = __importStar(require("fs"));
var TrackingMoreApi = /** @class */ (function () {
    function TrackingMoreApi(apiKey, baseURL) {
        var _this = this;
        if (baseURL === void 0) { baseURL = 'https://api.51tracking.com/v2'; }
        this.apiKey = apiKey;
        this.baseURL = baseURL;
        this.replaceUrl = function (url, data) {
            return data
                ? Object.entries(data).reduce(function (result, _a) {
                    var key = _a[0], value = _a[1];
                    return result.replace(RegExp("{" + key + "}", 'g'), value);
                }, url)
                : url;
        };
        this.qs = function (_a) {
            var method = _a.method, data = _a.data, url = _a.url;
            if (method === 'GET' && data && typeof data === 'object') {
                return url + "?" + qs.stringify(data);
            }
            return url;
        };
        this.request = function (url, method, data) {
            if (method === void 0) { method = 'GET'; }
            return rxjs_1.from(_this.axiosInstance.request({
                method: method,
                url: _this.qs({ method: method, url: url, data: data }),
                data: data,
            }));
        };
        this.requestFactory = function (url, method) {
            if (method === void 0) { method = 'GET'; }
            return function (data) { return _this.request(url, method, data); };
        };
        this.requestUrlFactory = function (url, method) {
            if (method === void 0) { method = 'GET'; }
            return function (select, data) {
                return _this.request(_this.replaceUrl(url, select), method, data);
            };
        };
        /**
         * 查询单个物流
         *
         * @memberof TrackingMoreApi
         */
        this.carriers = this.requestFactory('/carriers', 'GET');
        /**
         * 创建单个物流
         *
         * @memberof TrackingMoreApi
         */
        this.createTracking = this.requestFactory('/trackings/post', 'POST');
        /**
         * 更新物流信息
         *
         * @memberof TrackingMoreApi
         */
        this.updateTracking = this.requestUrlFactory('/trackings/{carrier_code}/{tracking_number}', 'PUT');
        /**
         * 删除订单
         *
         * @memberof TrackingMoreApi
         */
        this.deleteTracking = this.requestUrlFactory('/trackings/{carrier_code}/{tracking_number}', 'DELETE');
        /**
         * 获取单条物流
         *
         * @memberof TrackingMoreApi
         */
        this.getTracking = this.requestUrlFactory('/trackings/{carrier_code}/{tracking_number}', 'GET');
        /**
         * 批量添加物流
         *
         * @memberof TrackingMoreApi
         */
        this.batch = this.requestFactory('/trackings/batch', 'POST');
        /**
         * 实时查询，每秒3次
         *
         * @memberof TrackingMoreApi
         */
        this.realtime = this.requestFactory('/trackings/realtime', 'POST');
        /**
         * 更新物流公司
         *
         * @memberof TrackingMoreApi
         */
        this.update = this.requestFactory('/trackings/update', 'POST');
        /**
         * 获取各状态物流总数
         *
         * @memberof TrackingMoreApi
         */
        this.getStatusTotal = this.requestFactory('/trackings/getstatusnumber', 'GET');
        this.getList = function (parameter) {
            return _this.requestFactory('/trackings/get', 'GET')(TrackingMoreApi._toQueryString(parameter));
        };
        this.initCarriers = function () {
            return _this.carriers()
                .pipe(operators_1.pluck('data', 'data'), operators_1.map(function (items) {
                return items.reduce(function (prev, acc) {
                    prev[acc.code] = acc;
                    return prev;
                }, {});
            }), operators_1.tap(function (cmap) {
                fs.writeFileSync(__dirname + "/map.json", JSON.stringify(cmap));
            }))
                .subscribe();
        };
        this.axiosInstance = axios_1.default.create({
            baseURL: baseURL,
            headers: {
                'Tracking-Api-Key': apiKey,
                'Content-Type': 'application/json',
                Lang: 'cn',
            },
        });
        this.initCarriers();
    }
    TrackingMoreApi._toQueryString = function (parameter) {
        var result = __assign({}, parameter);
        if (parameter.orders) {
            result.orders = parameter.orders.join(',');
        }
        if (parameter.numbers) {
            result.numbers = parameter.numbers.join(',');
        }
        return result;
    };
    return TrackingMoreApi;
}());
exports.TrackingMoreApi = TrackingMoreApi;
