// @jsdvjx/query-fun/dist/nest/query-fun.module.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.QueryFunModule = void 0;
const tslib_1 = require("tslib");
const common_1 = require("@nestjs/common");
const index_1 = require("../index");
const remote_config_1 = require("@jsdvjx/remote-config");
const IQuerist_1 = require("../Queries/IQuerist");
let QueryFunModule = class QueryFunModule {
};
QueryFunModule = tslib_1.__decorate([
    common_1.Module({
        imports: [remote_config_1.ConfigModule],
        providers: [
            {
                provide: 'QueryFun',
                useFactory: async (rc) => {
                    //const rc = await RemoteConfig.create<any>(chunk.path);
                    const query = rc.map['query'];
                    const conf = query.databases.reduce((result, path) => {
                        const options = rc.map[path];
                        if (options instanceof Array) {
                            return result.concat(options);
                        }
                        else if (options.tag && options.type) {
                            result.push(options);
                        }
                        return result;
                    }, []);
                    for (const option of conf) {
                        await index_1.QueryFun.register(IQuerist_1.IQuerist.create(option));
                    }
                    return index_1.QueryFun.create(query.path, query.format);
                },
                inject: [remote_config_1.RemoteConfig],
            },
        ],
        exports: [index_1.QueryFun],
    })
], QueryFunModule);
exports.QueryFunModule = QueryFunModule;
