// sls-client/lib/client.module.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
var NestSlsClientModule_1;
"use strict";
const common_1 = require("@nestjs/common");
const client_1 = __importDefault(require("./client"));
let NestSlsClientModule = NestSlsClientModule_1 = class NestSlsClientModule {
    static register(option) {
        return {
            module: NestSlsClientModule_1,
            providers: [{
                    provide: "SlsClient",
                    useFactory: () => {
                        return new client_1.default(option);
                    }
                }],
            exports: [client_1.default]
        };
    }
};
NestSlsClientModule = NestSlsClientModule_1 = __decorate([
    common_1.Module({})
], NestSlsClientModule);
exports.default = NestSlsClientModule;
